src/main/scala/is/hail/methods/LDPrune.scala

```diff
@@ -0,0 +1,391 @@
+package is.hail.methods
+
+import java.util
+
+import is.hail.annotations.Annotation
+import is.hail.sparkextras.GeneralRDD
+import org.apache.spark.rdd.RDD
+import org.apache.spark.storage.StorageLevel
+import is.hail.sparkextras._
+import is.hail.variant._
+import is.hail.utils._
+
+object LDPrune {
+  val variantByteOverhead = 50
+  val fractionMemoryToUse = 0.25
+  val genotypesPerPack = 32
+  val nPartitionsPerCore = 3
+
+  case class GlobalPruneIntermediate(rdd: GeneralRDD[(Variant, BitPackedVector)], index: Int, persist: Boolean)
+
+  case class BitPackedVector(gs: Array[Long], nSamples: Int, mean: Double, sdRecip: Double) {
+    def unpack(): Array[Int] = {
+
+      val shifts = (0 until genotypesPerPack * 2 by 2).reverse
+
+      gs.flatMap { l =>
+        shifts.map { s => (l >> s) & 3 match {
+          case 0 => 0
+          case 1 => 1
+          case 2 => 2
+          case 3 => -1
+        }
+        }
+      }.take(nSamples)
+    }
+  }
+
+  val table = {
+    val t = Array.fill[Byte](256 * 4)(0)
+
+    (0 until 256).foreach { i =>
+      val xi = i & 3
+      val xj = (i >> 2) & 3
+      val yi = (i >> 4) & 3
+      val yj = (i >> 6) & 3
+
+      val res = findTableValue(xi, yi, xj, yj)
+
+      t(i * 4) = res._1.toByte
+      t(i * 4 + 1) = res._2.toByte
+      t(i * 4 + 2) = res._3.toByte
+      t(i * 4 + 3) = res._4.toByte
+    }
+    t
+  }
+
+  private def findTableValue(a: Int, b: Int, c: Int, d: Int): (Int, Int, Int, Int) = {
+    val r1 = findTableValue(a, b)
+    val r2 = findTableValue(c, d)
+    (r1._1 + r2._1, r1._2 + r2._2, r1._3 + r2._3, r1._4 + r2._4)
+  }
+
+  private def findTableValue(xi: Int, yi: Int): (Int, Int, Int, Int) = {
+    var xySum = 0
+    var XbarCount = 0
+    var YbarCount = 0
+    var XbarYbarCount = 0
+
+    (xi, yi) match {
+      case (3, 3) => XbarYbarCount += 1
```

Comment:
The left shift thing I was thinking of is:
```scala
(xi << 2) | yi match {
  case 0xf => ... // 0b1111   3,3
  case 0xd => ... // 0b1101   3,1
  case 0xe => ... // 0b1110   3,2
  case 0x7 => ... // 0b0111   1,3
  case 0xb => ... // 0b1011   2,3
  case 0xa => ... // 0b1010   2,2
  case 0x9 => ... // 0b1001   2,1
  case 0x6 => ... // 0b0110   1,2
  case 0x5 => ... // 0b0101   1,1
}
```

This is probably not worth the maintenance complexity. Both this and the above should compile to jump tables (i.e. use value to compute an offset and change current instruction to that offset, which takes constant-time in the number of cases).